File: src/main/java/web/database/DataBean.java
package web.database;

import jakarta.annotation.ManagedBean;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.faces.context.FacesContext;
import jakarta.inject.Named;

import web.models.PointModel;
import web.points.PointBean;

import java.io.FileReader;
import java.io.InputStream;
import java.net.URL;
import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Properties;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;

//@ManagedBean
@Named("dataBean")
@ApplicationScoped
public class DataBean {
    private Connection conn;
    private List<PointModel> points;


    public DataBean() {
        try {
            Class.forName("org.postgresql.Driver");
            Properties properties = new Properties();

            InputStream inputStream = getClass().getClassLoader().getResourceAsStream("db.cfg");
            properties.load(inputStream);
            inputStream.close();

            String user = properties.getProperty("user");
            String password = properties.getProperty("password");
            String url = properties.getProperty("URL");

            conn = DriverManager.getConnection(url, user, password);
        } catch (Exception e) {
            e.printStackTrace();
            return;
        }
    }


    public Connection getConn() {
        return conn;
    }

    public void setConn(Connection conn) {
        this.conn = conn;
    }

    public void addToTable(PointBean pointBean) {
        try(PreparedStatement statement = conn.prepareStatement("INSERT INTO webpoints(x, y, r, inArea) VALUES (?, ?, ?, ?);")) {
            statement.setDouble(1, pointBean.getX());
            statement.setDouble(2, pointBean.getY());
            statement.setDouble(3, pointBean.getR());
            statement.setBoolean(4, pointBean.isInArea());
            statement.execute();

        } catch (Exception e) {
            return;
        }
    }

    public void addToTableFromJS() {
        final Map<String, String> params = FacesContext.getCurrentInstance().getExternalContext().getRequestParameterMap();

        try(PreparedStatement statement = conn.prepareStatement("INSERT INTO webpoints(x, y, r, inArea) VALUES (?, ?, ?, ?);")) {
            PointBean p = new PointBean();

            p.setX(Double.parseDouble(params.get("x")));
            p.setY(Double.parseDouble(params.get("y")));
            p.setR(Double.parseDouble(params.get("r")));

            statement.setDouble(1, p.getX());
            statement.setDouble(2, p.getY());
            statement.setDouble(3, p.getR());
            statement.setBoolean(4, p.isInArea());
            statement.execute();

        } catch (Exception e) {
            return;
        }
    }

    public List<PointModel> getPoints() {
        points = new ArrayList<>();
        try (PreparedStatement statement = conn.prepareStatement("SELECT * FROM webpoints;");
             ResultSet resultSet = statement.executeQuery()) {

            while (resultSet.next()) {
                PointModel point = new PointModel();

                point.setX(resultSet.getDouble("x"));
                point.setY(resultSet.getDouble("y"));
                point.setR(resultSet.getDouble("r"));
                point.setInArea(resultSet.getBoolean("inArea"));
                point.setId(resultSet.getInt("attempt_id"));
                points.add(point);
            }

        } catch (SQLException e) {
            return null;
        }

        return points;
    }

    public void deletePoints() {
        try (PreparedStatement statement = conn.prepareStatement("TRUNCATE TABLE webpoints;")) {
            statement.execute();
        } catch (SQLException e) {
            return;
        }
    }

    public String getPointsJson() {
        try {
            ObjectMapper objectMapper = new ObjectMapper();
            return objectMapper.writeValueAsString(getPoints());
        } catch (JsonProcessingException e) {
            return "";
        }
    }
}

File: src/main/kotlin/web/database/DataBean.kt
package web.database

import com.fasterxml.jackson.databind.ObjectMapper
import jakarta.enterprise.context.ApplicationScoped
import jakarta.faces.context.FacesContext
import jakarta.inject.Named
import web.model.PointModel
import web.beans.PointBean
import java.sql.Connection
import java.sql.DriverManager
import java.sql.SQLException
import java.util.*

//@ManagedBean
@Named("dataBean")
@ApplicationScoped
class DataBean {
    var conn: Connection? = null
    //private var points: MutableList<PointModel> = mutableListOf()

    init {
        conn = try {
            Class.forName("org.postgresql.Driver")
            val properties = Properties()
            val inputStream = javaClass.getClassLoader().getResourceAsStream("db.cfg")
            properties.load(inputStream)
            inputStream.close()
            val user = properties.getProperty("user")
            val password = properties.getProperty("password")
            val url = properties.getProperty("URL")
            DriverManager.getConnection(url, user, password)
        } catch (e: Exception) {
           error(e)
        }
    }


    fun addToTable(pointBean: PointBean) {
        val sql = "INSERT INTO webpoints(x, y, r, inArea) VALUES (?, ?, ?, ?);"
        executeUpdate(sql) { statement ->
            statement.setDouble(1, pointBean.x)
            statement.setDouble(2, pointBean.y)
            statement.setDouble(3, pointBean.r)
            statement.setBoolean(4, pointBean.isInArea)
        }
    }

    fun addToTableFromJS() {
        val params = FacesContext.getCurrentInstance().externalContext.requestParameterMap
        val pointBean = PointBean().apply {
            x = params["x"]?.toDouble() ?: 0.0
            y = params["y"]?.toDouble() ?: 0.0
            r = params["r"]?.toDouble() ?: 0.0
        }

        addToTable(pointBean)
    }

    fun getPoints(): List<PointModel> {
        val sql = "SELECT * FROM webpoints;"
        return executeQuery(sql) { resultSet ->
            generateSequence { if (resultSet.next()) resultSet else null }.map {
                PointModel().apply {
                    x = it.getDouble("x")
                    y = it.getDouble("y")
                    r = it.getDouble("r")
                    isInArea = it.getBoolean("inArea")
                    id = it.getInt("attempt_id")
                }
            }.toList()
        }
    }

    fun deletePoints() {
        val sql = "TRUNCATE TABLE webpoints;"
        executeUpdate(sql)
    }


    private fun <T> executeQuery(sql: String, handler: (java.sql.ResultSet) -> T): T {
        conn?.let { connection ->
            connection.prepareStatement(sql).use { statement ->
                statement.executeQuery().use { resultSet ->
                    return handler(resultSet)
                }
            }
        } ?: throw SQLException("Database connection is not initialized")
    }

    private fun executeUpdate(sql: String, statementHandler: (java.sql.PreparedStatement) -> Unit = {}) {
        conn?.let { connection ->
            connection.prepareStatement(sql).use { statement ->
                statementHandler(statement)
                statement.executeUpdate()
            }
        } ?: throw SQLException("Database connection is not initialized")
    }


    val pointsJson: String
        get() = ObjectMapper().writeValueAsString(getPoints())

}
